File: ios/Touliao/Features/Search/SearchFilters.swift
import Foundation

// MARK: - 全局搜索分类筛选（F5，对齐 Web utils/messageSearchFilters.js / Android SearchFilters.kt：
// type 取值/时间区间语义/摘要规则四端口径统一）

/// 类型筛选选项（value 空串 = 全部；映射后端 messages.type）
struct SearchTypeOption: Identifiable, Equatable {
    let value: String
    let label: String
    let icon: String
    var id: String { value }
}

let messageSearchTypes: [SearchTypeOption] = [
    .init(value: "", label: "全部", icon: "○"),
    .init(value: "text", label: "文本", icon: "文"),
    .init(value: "image", label: "图片", icon: "▧"),
    .init(value: "voice", label: "语音", icon: "◖"),
    .init(value: "video", label: "视频", icon: "▶"),
    .init(value: "file", label: "文件", icon: "▤"),
    .init(value: "contact_card", label: "名片", icon: "人"),
    .init(value: "red_packet", label: "红包", icon: "包"),
    .init(value: "transfer", label: "转账", icon: "¥"),
    .init(value: "merged", label: "合并转发", icon: "☷"),
    .init(value: "call", label: "通话", icon: "☎"),
]

/// 时间筛选：空串=不限 | today=今天 | 7d=近7天 | 30d=近30天（秒级 from/to）
struct SearchTimeRangeOption: Identifiable, Equatable {
    let value: String
    let label: String
    var id: String { value.isEmpty ? "any" : value }
}

let messageSearchTimeRanges: [SearchTimeRangeOption] = [
    .init(value: "", label: "不限"),
    .init(value: "today", label: "今天"),
    .init(value: "7d", label: "7天"),
    .init(value: "30d", label: "30天"),
]

/// 由筛选状态构造请求参数（与 Web buildMessageSearchParams 同口径：
/// today → 今天 0 点起；7d/30d → 从当天 0 点再往前推 N 天）。
func buildSearchFilterParams(type: String, timeRange: String, senderId: String,
                             now: Date = Date()) -> SearchFilterParams {
    if type.isEmpty && timeRange.isEmpty && senderId.isEmpty { return SearchFilterParams() }
    var params = SearchFilterParams(
        type: type.isEmpty ? nil : type,
        senderId: senderId.isEmpty ? nil : senderId
    )
    if !timeRange.isEmpty {
        let cal = Calendar.current
        var start = cal.startOfDay(for: now)
        if timeRange == "7d" { start = cal.date(byAdding: .day, value: -7, to: start) ?? start }
        if timeRange == "30d" { start = cal.date(byAdding: .day, value: -30, to: start) ?? start }
        params.fromSec = Int64(start.timeIntervalSince1970)
        params.toSec = Int64(now.timeIntervalSince1970)
    }
    return params
}

/// 结果行类型图标
func messageSearchTypeIcon(_ type: String) -> String {
    messageSearchTypes.first { $0.value == type }?.icon ?? "•"
}

/// 解析结构化消息 content（名片/红包/转账/合并转发 的 JSON 字段），坏 JSON 返回空字典
private func parseContentObject(_ content: String) -> [String: Any] {
    guard let data = content.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return [:] }
    return obj
}

private func compact(_ label: String, _ detail: String = "") -> String {
    let trimmed = detail.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? "[\(label)]" : "[\(label)] \(trimmed)"
}

/// 结果摘要（对齐 Web formatSearchMessageSummary / Android formatSearchMessageSummary）：
/// 结构化消息只透出人话字段（名片备注/红包祝福/转账备注/合并标题），不泄原始 JSON。
func formatSearchMessageSummary(type: String, content: String) -> String {
    let body = content.trimmingCharacters(in: .whitespacesAndNewlines)
    switch type {
    case "text": return body
    case "image": return compact("图片", body)
    case "voice": return compact("语音")
    case "video": return compact("视频", body)
    case "file": return compact("文件", body)
    case "contact_card", "contact": {
        let obj = parseContentObject(body)
        let name = (obj["remark"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            ?? (obj["username"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            ?? (obj["name"] as? String) ?? ""
        return compact("名片", name)
    }()
    case "red_packet": return compact("红包", (parseContentObject(body)["greeting"] as? String) ?? "")
    case "transfer": return compact("转账", (parseContentObject(body)["note"] as? String) ?? "")
    case "merged": return compact("聊天记录", (parseContentObject(body)["title"] as? String) ?? "")
    case "call": return compact("通话", body)
    default: return body.isEmpty ? compact(type) : body
    }
}

File: android/app/src/main/java/com/touliao/app/feature/search/SearchFilters.kt
package com.touliao.app.feature.search

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.jsonObject

/**
 * 全局搜索分类筛选（F4b，对齐 Web utils/messageSearchFilters.js：
 * type 取值/时间区间语义/摘要规则与 Web 完全一致，四端口径统一）。
 */

/** 类型筛选选项（value 空串 = 全部；映射后端 messages.type） */
data class SearchTypeOption(val value: String, val label: String, val icon: String)

val MESSAGE_SEARCH_TYPES: List<SearchTypeOption> = listOf(
    SearchTypeOption("", "全部", "○"),
    SearchTypeOption("text", "文本", "文"),
    SearchTypeOption("image", "图片", "▧"),
    SearchTypeOption("voice", "语音", "◖"),
    SearchTypeOption("video", "视频", "▶"),
    SearchTypeOption("file", "文件", "▤"),
    SearchTypeOption("contact_card", "名片", "人"),
    SearchTypeOption("red_packet", "红包", "包"),
    SearchTypeOption("transfer", "转账", "¥"),
    SearchTypeOption("merged", "合并转发", "☷"),
    SearchTypeOption("call", "通话", "☎"),
)

/** 时间筛选：空串=不限 | today=今天 | 7d=近7天 | 30d=近30天（秒级 from/to） */
data class SearchTimeRangeOption(val value: String, val label: String)

val MESSAGE_SEARCH_TIME_RANGES: List<SearchTimeRangeOption> = listOf(
    SearchTimeRangeOption("", "不限"),
    SearchTimeRangeOption("today", "今天"),
    SearchTimeRangeOption("7d", "7天"),
    SearchTimeRangeOption("30d", "30天"),
)

/** 搜索请求参数（不含 q/limit；null 字段不随请求发送） */
data class SearchFilterParams(
    val type: String? = null,
    val fromSec: Long? = null,
    val toSec: Long? = null,
    val senderId: String? = null,
) {
    val isEmpty: Boolean get() = type == null && fromSec == null && toSec == null && senderId == null
}

private fun startOfDay(epochMillis: Long): Long {
    val cal = java.util.Calendar.getInstance().apply { timeInMillis = epochMillis }
    cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
    cal.set(java.util.Calendar.MINUTE, 0)
    cal.set(java.util.Calendar.SECOND, 0)
    cal.set(java.util.Calendar.MILLISECOND, 0)
    return cal.timeInMillis
}

/**
 * 由筛选状态构造请求参数。
 * today → 今天 0 点起；7d/30d → 从当天 0 点再往前推 N 天（与 Web buildMessageSearchParams 同口径）。
 * @param nowMillis 当前时间（毫秒），测试注入固定值
 */
fun buildSearchFilterParams(
    type: String,
    timeRange: String,
    senderId: String,
    nowMillis: Long = System.currentTimeMillis(),
): SearchFilterParams {
    if (type.isBlank() && timeRange.isBlank() && senderId.isBlank()) return SearchFilterParams()
    var params = SearchFilterParams(
        type = type.takeIf { it.isNotBlank() },
        senderId = senderId.takeIf { it.isNotBlank() },
    )
    if (timeRange.isNotBlank()) {
        var start = startOfDay(nowMillis)
        if (timeRange == "7d") start -= 7L * 24 * 3600 * 1000
        if (timeRange == "30d") start -= 30L * 24 * 3600 * 1000
        params = params.copy(fromSec = start / 1000, toSec = nowMillis / 1000)
    }
    return params
}

private fun searchTypeIconOf(type: String): String =
    MESSAGE_SEARCH_TYPES.firstOrNull { it.value == type }?.icon ?: "•"

/** 结果行类型图标 */
fun messageSearchTypeIcon(type: String): String = searchTypeIconOf(type)

private fun searchTypeLabel(type: String): String =
    MESSAGE_SEARCH_TYPES.firstOrNull { it.value == type }?.label ?: type

private val summaryJson = Json { ignoreUnknownKeys = true }

/** 解析结构化消息 content（名片/红包/转账/合并转发 的 JSON 字段），坏 JSON 返回空对象 */
private fun parseContentObject(content: String): JsonObject =
    runCatching { summaryJson.parseToJsonElement(content).jsonObject }.getOrNull() as? JsonObject ?: JsonObject(emptyMap())

private fun JsonObject.stringOf(key: String): String =
    runCatching { get(key)?.jsonPrimitive?.content }.getOrNull().orEmpty()

private fun compact(label: String, detail: String = ""): String {
    val safe = detail.trim()
    return if (safe.isNotEmpty()) "[$label] $safe" else "[$label]"
}

/**
 * 结果摘要（对齐 Web formatSearchMessageSummary）：
 * 结构化消息只透出人话字段（名片备注/红包祝福/转账备注/合并标题），不泄原始 JSON。
 */
fun formatSearchMessageSummary(type: String, content: String): String {
    val body = content.trim()
    return when (type) {
        "text" -> body
        "image" -> compact("图片", body)
        "voice" -> compact("语音")
        "video" -> compact("视频", body)
        "file" -> compact("文件", body)
        "contact_card" -> {
            val obj = parseContentObject(body)
            compact("名片", obj.stringOf("remark").ifBlank { obj.stringOf("username") }.ifBlank { obj.stringOf("name") })
        }
        "red_packet" -> compact("红包", parseContentObject(body).stringOf("greeting"))
        "transfer" -> compact("转账", parseContentObject(body).stringOf("note"))
        "merged" -> compact("聊天记录", parseContentObject(body).stringOf("title"))
        "call" -> compact("通话", body)
        else -> body.ifBlank { compact(searchTypeLabel(type)) }
    }
}
